src/tpwm.c

```diff
@@ -0,0 +1,122 @@
+/*
+ * tpwm.c - Uncomprassing TPWM'ed content
+ *
+ * Copyright (C) 2015  Wicked_Digger  <wicked_digger@mail.ru>
+ *
+ * This file is part of freeserf.
+ *
+ * freeserf is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * freeserf is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with freeserf.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "tpwm.h"
+
+#include <stdlib.h>
+#include <string.h>
+
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#ifdef HAVE_STDINT_H
+# include <stdint.h>
+#endif
+
+static const uint8_t tpwm_sign[4] = {'T','P','W','M'};
+
+#pragma pack(push,1)
+typedef struct {
+	uint8_t magic[4];
+	uint32_t original_size;
+} tpwm_header_t;
+#pragma pack(pop)
+
+int
+tpwm_is_compressed(void *src_data, unsigned int src_size)
+{
+	if (src_size < 8) {
+		return -1;
+	}
+
+	tpwm_header_t *header = (tpwm_header_t*)src_data;
+	if (0 != memcmp(header->magic, tpwm_sign, 4)) {
+		return -1;
+	}
+
+	return 0;
+}
+
+int
+tpwm_uncompress(void *src_data, unsigned int src_size,
+                void **res_data, unsigned int *res_size, char **error)
+{
+	if ((res_data == NULL) || (res_size == NULL)) {
+		*error = "TPWM: bad parameter";
+		return -1;
+	}
+
+	if (tpwm_is_compressed(src_data, src_size) < 0) {
+		*error = "TPWM: source buffer is not tpwm packed";
+		return -1;
+	}
+
+	*res_data = NULL;
+	*res_size = 0;
+
+	tpwm_header_t *header = (tpwm_header_t*)src_data;
+
+	*res_data = malloc(header->original_size);
```

Comment:
You have to convert the `original_size` field from little-endian 32bit to native (using `le32toh`). I would simply do `le32toh(*(uint32_t *)&src_data[4])`.